### ISSUE TITLE
Persistent templates

### ISSUE THREAD

**Issue by jaragunde** [Igalia/phpreport#254]
Currently templates are saved as cookies in the browser, so they are lost if the user uses a different browser for some reason, or in case the PhpReport deployment changes its URL.

It would be a good idea to make templates persistent in the DB. This would imply creating web services to add, remove and get them from the DB and, of course, implement the necessary objects in the model layers.

We would also need a page to migrate the cookie-based templates to the DB. Users would have to use it once before being able to use their templates again in an upgraded PhpReport.

**Comment by jaragunde** [comment]
Preliminary implementation pushed to branch `persistent-templates`.

**Comment by jaragunde** [comment]
Preliminary implementation bypasses the backend layers structure; it should be addressed next.

**Comment by jaragunde** [comment]
I found a bug in the ongoing implementation. To reproduce it:
* Create a new task, check the "onsite" box, then call it "onsite task" and create a template for it.
* Click on the newly created template button called "onsite task". See the checkbox for the new task is not checked.
* If you refresh the page, the new template works properly, which proves it was stored properly but there is a problem when templates are fresh.

The same happens with the "telework" box.

**Comment by jaragunde** [comment]
@tonythomas01 found another bug when creating templates with no name. These should be addressed before merging the feature into master.

**Comment by tonythomas01** [comment]
> Click on the newly created template button called "onsite task". See the checkbox for the new task is not checked.

addressed in https://github.com/Igalia/phpreport/pull/270/commits/1df8c0e7977572dd25058efac15c42d81af5b901 

> @tonythomas01 found another bug when creating templates with no name. These should be addressed before merging the feature into master.

addressed in https://github.com/Igalia/phpreport/pull/270/commits/30d414e52e8261cb7539d82618d6f23d1aa7bc3c

both are added to the PR at https://github.com/Igalia/phpreport/pull/270

**Comment by jaragunde** [comment]
```
commit bce20fa080c7d37d59719575f73128cc90847811
Merge: 77a1210 342de96
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Wed Jul 27 12:04:19 2016 +0200

    [#254] Implement persistent templates.
    
    This patch set implements persistent templates, a page to migrate
    cookie-based templates to persistent ones, and also fixes #265.
    
    (merge branch 'persistent-templates')

commit 342de9669c144766d2f28205d4a0c6e21642a0e9
Author: Tony Thomas <01tonythomas@gmail.com>
Date:   Wed Jul 27 15:18:19 2016 +0530

    Update templates documentation to add new 'template name' feature

commit 9e3bd9b0166e6042560eeaed8f95e576b3555485
Author: Tony Thomas <01tonythomas@gmail.com>
Date:   Wed Jul 27 15:01:39 2016 +0530

    [#265] Include description of the task too in the template

commit 1df8c0e7977572dd25058efac15c42d81af5b901
Author: Tony Thomas <01tonythomas@gmail.com>
Date:   Wed Jul 27 10:32:52 2016 +0530

    [#254] Make sure the bool fields are set correctly for a fresh template load

commit 30d414e52e8261cb7539d82618d6f23d1aa7bc3c
Author: Tony Thomas <01tonythomas@gmail.com>
Date:   Tue Jul 26 22:30:44 2016 +0530

    [#265] Separately ask for template name using Ext message prompt

commit d9dd8a7cb2657e98a1f974ff9ffc84ce28666d47
Author: Tony Thomas <01tonythomas@gmail.com>
Date:   Sun Jul 24 22:15:05 2016 +0530

    [#254] Make all template actions comply with phpreport DB abstractions

commit de3acba1a72fb7a19a48f2c31941f39a2a46e9ba
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Tue Jul 19 17:30:22 2016 +0200

    Check result of migrate operation and remove cookies if done.
    
    The Store in migrate-templates.php has been changed so it can read the
    response to the creation operation, and remove the cookie if it was
    successful.

commit c271f3931a45e926ca5b2407039da9ed5e9f24ec
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Tue Jul 19 14:12:54 2016 +0200

    Add foreign key constraints to the templates table.

commit f75f22b85a9b0f2172f14f3f9db5b829fb5aba30
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Tue Jul 19 13:46:07 2016 +0200

    Implement persistent task deletion from the tasks screen.
    
    Created a service to delete tasks.
    
    FIXME: the service contains hard-coded DB queries, violating the
    design of the model layer.
    
    FIXME: the service may remove tasks that belong to different users.

commit ee6cb4196c1c3dcbf15b08043f7f23d352befacf
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Tue Jul 19 13:20:37 2016 +0200

    Implement persistent task creation from the tasks screen.
    
    Now the button to create templates in the tasks screen adds them to
    a store, which is configured to create them in the backend via the
    createTemplatesService.
    
    The service required some changes to return the list of created tasks
    with their ids generated by the database.

commit 8934f87a7ba52363689de0e929dc6b758c5b424a
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Mon Jul 18 16:30:17 2016 +0200

    Preliminary implementation of template retrieval from DB.
    
    Create a service to read the templates and modify the tasks screen to
    read tasks from the service instead of the cookies.
    
    FIXME: the service contains hard-coded DB queries, violating the
    design of the model layer.

commit 7176a2fbf630efcf4f2c2e275e732ad4b5b043f0
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Mon Jul 18 16:21:07 2016 +0200

    Preliminary implementation of persistent templates creation.
    
    This patch contains the creation service and the page to migrate the
    cookie based templates to persistent templates.
    
    FIXME: the creation service contains hard-coded DB queries, violating
    the design of the model layer.
    
    FIXME: the migration page does not check if the migration completed
    successfully and doesn't remove the templates cookie.
```

**Comment by jaragunde** [comment]
Added an extra commit to fix template migration in the specific case that users are upgrading from 2.0:
```
commit 2a16af6e13092a341cd2b3b93b503ede02d81dec
Author: Jacobo Aragunde Pérez <jaragunde@igalia.com>
Date:   Tue May 30 13:16:29 2017 +0200

    [#254] Allow migration of templates created in 2.0
    
    These templates had one field less, because onsite was added in 2.1.

```
